### PR TITLE
command_output: do not treat text on stderr as failure, rely on exit code

### PR DESCRIPTION
**Pull request by maximbaz**
Often scripts put some diagnostic information on stderr, so having some text on stderr doesn't necessarily mean the command has failed. In fact, this is true in most cases — exit code is an indicator of whether command failed or not.

This PR allows to ignore stderr and thus not failing the command. This PR also changes the default behavior of `external_script` because I believe it is more correct this way.

Not sure on the name of the property. Maybe reversing is better, like `fail_on_stderr`?

### CODE REVIEW

**Comment by maximbaz** [comment]
Travis seems to be broken 😕

**Comment by lasers** [comment]
@ultrabug Can you talk with Travis? He's failing my PRs too. Ask Travis why.

File: py3status/modules/external_script.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
     cache_timeout: how often we refresh this module in seconds
         (default 15)
     format: see placeholders below (default '{output}')
+    ignore_stderr: do not treat output on stderr as failure
+        (defaut True)
```
Comment (lasers) [inline py3status/modules/external_script.py]: Hi, I'm your Travis today. What is `defaut`? I do not know if it is good idea to leave this on by `defaut`.

File: py3status/py3.py [hunk 2]
```diff
@@ -933,7 +933,7 @@ def command_run(self, command):
             msg = 'Command `{cmd}` {error}'.format(cmd=pretty_cmd, error=e.errno)
             raise exceptions.CommandError(msg, error_code=e.errno)
 
-    def command_output(self, command, shell=False, capture_stderr=False, localized=False):
+    def command_output(self, command, shell=False, capture_stderr=False, localized=False, ignore_stderr=False):
```
Comment (lasers) [inline py3status/py3.py]: In many modules, we use this.
```python
        try:
            self.py3.command_output('REPLACE_ME')
        except self.py3.CommandError as ce:
            return ce.output
```

File: py3status/modules/external_script.py (hunk 1 above)
Comment (maximbaz) [reply py3status/modules/external_script.py]: As I mentioned in the first message, I don't think the current behavior in master makes any sense, very often scripts print some additional information on stderr and it doesn't mean the script failed. Even black formatter prints some emojis on stderr upon successful completion

File: py3status/py3.py (hunk 2 above)
Comment (maximbaz) [reply py3status/py3.py]: To be honest, maybe the default here should be changed to True as well.

I'm not sure what you wanted to showcase with your example?

Comment (lasers) [reply py3status/py3.py]: An example above is a clear indicator of an `out to stderr` text.

If we use `ignore_stderr=False`, we can't tell easily which std throws an output. In the long run, we will end up assuming it's always `stderr` rather than seeing it can be cycled between both.

Also, per module error messages sometimes have to be fixed first (eg removing prefix, omitting empty lines, or picking certain line, etc). If we do have to fix them, then what is the point of having this argument in first place?

Simply put, you could be adding an argument that nobody would use... And it's more about having an ability to handle the output better in all modules than to simplify something inside `py3` so you can kill off few lines in your new module.

Comment (lasers) [reply py3status/py3.py]: See `arch_updates` module. Right now, it uses `subprocess.CalledProcessError` and should be converted to `self.py3.command_output()` eventually. Anyway, we need `try/except` here so we can return `None` when users can't get an output.

Users can't get an output because they're not connected to Internet. With `None`, users can make their `arch_updates` disappear (eg `[{pacman}/{aur}]`) or display something else.

Comment (lasers) [reply py3status/py3.py]: `external_script` appears to not have something to deal with `stderr` outputs.
```diff
diff --git a/py3status/modules/external_script.py b/py3status/modules/external_script.py
index 6bc231cb..969e0c82 100644
--- a/py3status/modules/external_script.py
+++ b/py3status/modules/external_script.py
@@ -54,6 +54,7 @@ class Py3status:
     localize = True
     script_path = None
     strip_output = False
+    ignore_stderr = False
 
     def post_config_hook(self):
         if not self.script_path:
@@ -73,7 +74,8 @@ class Py3status:
         except self.py3.CommandError as e:
             # something went wrong show error to user
             output = e.output or e.error
-            self.py3.error(output)
+            if not self.ignore_stderr:
+                self.py3.error(output)
 
         if output_lines:
             output_text = output_lines[0]

```

Comment (lasers) [reply py3status/py3.py]: Repeat for this conversation: I did something [bad](https://github.com/lasers/py3status/tree/newmodule-external_script). First draft.

Comment (maximbaz) [reply py3status/py3.py]: Maybe I'm still missing the point, but the patch above has one flaw: it ignores ALL errors, even true errors. I do want to execute self.py3.error() if command exited with a non-zero exit code, because non-zero exit code is what normally indicates an error. But if command exited with zero exit code, then I don't want to execute self.py3.error. 

I'll have another look at arch_updates and your new module in the evening, I'm currently replying from phone.

Comment (maximbaz) [reply py3status/py3.py]: @lasers I'm looking at `arch_updates`, that case is special because `cower` is broken`, `cower --update` returns non-zero status code on success. This is abnormal. We can change the command to be `cower --update || true`, this will always return zero exit code and so we can process the output as normal.

**Comment by maximbaz** [comment]
You know, I actually think we don't need a new parameter at all, as I think there is no case when we ever need to treat command as failed if there is some text on stderr. Am happy to be proven wrong with some real-world counter-examples 🙂

**Comment by maximbaz** [comment]
I think you are missing the point, let me try to expand on your example. 

Let's say 'get_data()' executes a script that prints the following: 

On stdout: name=X, address=Y, zipcode=Z
On stderr: data is successfully retrieved!🤘
Exit code: 0

This is precisely what 'black' formatter is doing. Even though there is some text on stderr, the script execution was successful, so we should not go into 'except' and should not stop the script.

Now let's take a different case, where the script failed: 

On stdout: 
On stderr: execution failed because of missing dependency 
Exit code: 1

In this case non-zero exit code indicates a failure, now I do want to go inside 'except', stop parsing the data and show an error to the user.


In master, both examples result in user error.
In this PR, the first case is considered a success (as it should be), and only the second case is handled as an error. 

Does this clarify what I'm trying to achieve?

**Comment by lasers** [comment]
>Does this clarify what I'm trying to achieve?

Look at `newmodule-external_script` again. I went overboard. Test it. Make modifications. It should do what you want... or almost to it.

**Comment by maximbaz** [comment]
Yes, as I wrote in another PR I agree that your update to execute_script is better than me creating a new module. Do you want to make a PR with it, or do you want me to copy your code and submit a PR myself?

To clarify, this PR is still needed, it is impossible to distinguish the two examples I showed above from a module, only this PR can fix this issue.

**Comment by lasers** [comment]
I don't want to make that PR yet because I hadn't tested it thoroughly and I still need to work on that area. Also, I need to iterate this with other contributors. I want to see if I can add `Popen` support so we don't have to merge `async_script` #1265  too.

**Comment by maximbaz** [comment]
If you want, I can take over your current code, test it and submit a PR that fixes my use-case, which will make cached_script proposal obsolete - and we can tackle Popen another time. If this helps speeding up things of course, I simply want to get a satisfying solution merged 😋

**Comment by lasers** [comment]
Sure. Test it. Submit a minimal `external_script` PR that fixes your use-case. `external_script` is just missing error controls which shouldn't be too hard to implement after looking at my code.

**Comment by tobes** [comment]
Ok, My take is that this simplifies the code and shouldn't really break anything so let's just go for it.
